torch/testing/_comparison.py

```diff
@@ -0,0 +1,986 @@
+import abc
+import cmath
+import collections.abc
+from typing import NamedTuple, Callable, Sequence, List, Union, Optional, Type, Tuple, Any, cast
+
+import numpy as np
```

Comment:
We do require NumPy to run our tests, but torch.testing is a public module that's imported by default:

https://github.com/pytorch/pytorch/blob/191b48b12f33e1e9525882da0c62b68686d69e42/torch/__init__.py#L748

So yes, including NumPy here will make NumPy a dependency of PyTorch. 

We could require NumPy to use torch.testing, but maybe it's easier to just guard the NumPy references here to be conditional on the NumPy module being imported? Creating base classes that are extended if NumPy is available seems like the same solution, so however you want to implement it sounds good.

Reply:
Just to be sure: we do want to have the capability for the `numpy` stuff in the public testing side, i.e. `assert_close`, if they are available, correct? That would extend our current functionality. Note that we are only talking about `numpy` scalars here, which are something different than 0d arrays (I know, weird). `numpy.ndarrays` are supported anyway.

I'm asking, because I could also live with that not being on the public side and only available for internal stuff.

Reply:
It'd be nice to have the NumPy-specific logic in the public torch.testing module if a user is using NumPy. If it's a giant pain to support I could also see doing without.